android/src/main/java/com/ioreactnativecieid/IoReactNativeCieidModule.kt

```diff
@@ -28,9 +29,40 @@ class IoReactNativeCieidModule(reactContext: ReactApplicationContext) :
 
   override fun getName() = NAME
 
+  private fun isSignatureValid(packageName: String, signature: String): Boolean {
+    val pm = reactApplicationContext.packageManager
+    val packageInfo = if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.P) {
+      pm.getPackageInfo(packageName, PackageManager.GET_SIGNING_CERTIFICATES)
+    } else {
+      @Suppress("DEPRECATION")
+      pm.getPackageInfo(packageName, PackageManager.GET_SIGNATURES)
+    }
+
+    val signatures = if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.P) {
+      packageInfo.signingInfo.apkContentsSigners
+    } else {
+      @Suppress("DEPRECATION")
+      packageInfo.signatures
+    }
+
+    val sha256List = signatures.map {
+      val digest = MessageDigest.getInstance("SHA-256")
+      digest.update(it.toByteArray())
+      // Converts bytes in hexadecimal uppercase string
+      digest.digest().joinToString(":") { byte ->
+        "%02x".format(byte).uppercase()
+      }
+    }
+    // Returns the first sha256 signature
+    return sha256List.contains(signature)
+  }
+
   @ReactMethod(isBlockingSynchronousMethod = true)
-  fun isAppInstalled(packageName: String): Boolean = try {
+  fun isAppInstalled(packageName: String, signature: String): Boolean = try {
     reactApplicationContext.packageManager.getPackageInfo(packageName, 0)
+    if(!isSignatureValid(packageName, signature)) {
+      throw PackageManager.NameNotFoundException()
+    }
     true
   } catch (e: PackageManager.NameNotFoundException) {
     false
```

Comment:
```suggestion
  @ReactMethod(isBlockingSynchronousMethod = true)
  fun isAppInstalled(packageName: String, signature: String): Boolean = try {
    reactApplicationContext.packageManager.getPackageInfo(packageName, 0)
    runCatching { isSignatureValid(packageName, signature) }.getOrDefault(false)
  } catch (e: PackageManager.NameNotFoundException) {
    false
```
What about this version?
- We don't tie the meaning of the `NameNotFoundException` exception to a signature mismatch.
- If for some reason the method isSignatureValid throws an exception different from `NameNotFoundException ` we don't crash the app.

What do you think?

Reply:
Great suggestion! I think the solution you suggest is more correct and more elegant than mine, thanks!